GPII-3132: Rollup grunt linting plugin.

the-t-in-rtf opened the pull request:
See [GPII-3132](https://issues.gpii.net/browse/GPII-3132) for details.

the-t-in-rtf committed:
GPII-3132: Initial implementation

the-t-in-rtf committed:
GPII-3132: Initial implementation.

the-t-in-rtf committed:
GPII-3132: Reverted version from aborted fluid-publish run.

the-t-in-rtf committed:
GPII-3132: Fixed package name.

amb26 commented on README.md:
````diff
@@ -1,2 +1,68 @@
+command like:
+
+```shell
+npm install grunt-gpii-grunt-lintall --save-dev
````
should be lint-all, and throughout here

the-t-in-rtf replied:
Done.

amb26 commented on index.js:
```diff
@@ -0,0 +1,17 @@
+var fluid = require("infusion");
+
+// Register this module so that people can reuse our configuration, etc. using `fluid.module.resolvePath("%gpii-grunt-lintall/path/to/content")`.
+fluid.module.register("gpii-grunt-lintall", __dirname, require);
```
This should use the module's name

the-t-in-rtf replied:
Fixed, throughout.

amb26 commented on tasks/lintall.js:
```diff
@@ -0,0 +1,53 @@
+    grunt.loadNpmTasks("grunt-markdownlint");
+
+    // By default, lint and run all tests.
+    grunt.registerTask("lintall", "Apply eslint, jsonlint, json5lint, and various markdown linting checks", ["eslint:js", "jsonlint", "json5lint", "markdownlint", "eslint:md", "mdjsonlint"]);
```
lint-all?

the-t-in-rtf replied:
Done.

the-t-in-rtf committed:
GPII-3132: Standardised package name and task name throughout.

the-t-in-rtf commented:
@amb26, I have standardised the package and task name throughout, this should be ready for further review.

amb26 commented on .eslintrc-md.json:
```diff
@@ -0,0 +1,15 @@
+    "plugins": [
+        "markdown"
+    ]
+}
```
Suggestion (feel free to ignore) that we adopt a further linting plugin such as https://github.com/schorfES/grunt-lintspaces to enforce final newlines in this project and others

the-t-in-rtf replied:
Done and tested in depth with universal.  I ended up having to disable their indentation linting for Javascript comments and markdown, but otherwise it's hugely useful in catching syntax problems in previous unlinted files.

amb26 commented on Gruntfile.js:
```diff
@@ -0,0 +1,40 @@
+/*
+ * grunt-gpii-grunt-lint-all
```
Extra "grunt" here

amb26 commented on LICENSE-BSD-3-Clause:
```diff
@@ -0,0 +1,27 @@
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+
+* Neither the name of the gpii-grunt-lint-all nor the names of its
```
It seems hard to imagine this name being used to endorse or promote anything :)

the-t-in-rtf replied:
You caught me, that bit of boilerplate game from `grunt init`.  Removed in favour of a modified version of the license in universal (updated to 2018, minus the funding wording).

the-t-in-rtf committed:
GPII-3132: Centralised settings across all checks.  Added more complete JSON(5) linting.

the-t-in-rtf committed:
GPII-3132: Aliased standard 'lint' command for use with prehooks based on our standard task names.

the-t-in-rtf committed:
GPII-3132: Added grunt-lintspaces to rollup and fixed linting errors revealed.

the-t-in-rtf committed:
GPII-3132: Removed duplicate and faulty indentation check from eslint-json task.

the-t-in-rtf committed:
GPII-3132: Updated json-eslint rules to allow ES5 multi-line strings.

the-t-in-rtf committed:
GPII-3132: Excepted js-comments from lintspaces indentation checking.

the-t-in-rtf committed:
GPII-3132: Replaced boilerplate license generated by Grunt with something closer to our standard.

the-t-in-rtf commented:
@amb26, this should be ready for review again, or perhaps you can look at the universal pull to see what changes the harness ends up requiring practically.

amb26 commented:
The mdjsonlint and json-eslint tasks seem to have a reporting fault in that they don't report a file count - from a current universal run:

````
Running "mdjsonlint:src" (mdjsonlint) task
>> files lint free.

Running "json-eslint:src" (json-eslint) task
>> files lint free.
````

amb26 commented:
Other than that, all this work looks brilliant and we have just the minimal configuration and boilerplate required that we would want

amb26 commented on .eslintrc.json:
```diff
@@ -0,0 +1,6 @@
+    "env": {
+        "node": true
+    }
+}
```
This project should be linted with respect to itself - so that these newlines get added

amb26 replied:
I guess it already is - and it's just a matter of adding these dotfiles to the JSON task

the-t-in-rtf replied:
Exactly, they just weren't checked.  I added a rule to cover dot files in the root and fixed the errors revealed.  I also manually tidied up the files not covered by any checks, i.e. the LICENSE.txt and .gitignore.

the-t-in-rtf replied:
I realised that .gitignore and other top-level dot files were common enough that I should add an "other" category that's only used with the carriage return check.

amb26 commented on tasks/json-eslint.js:
```diff
@@ -0,0 +1,69 @@
+            fluid.each(validPaths, function (filepath) {
+                try {
+                    var jsonContent       = grunt.file.read(filepath);
+                    var wrappedContent    = fluid.stringTemplate("/* eslint-env node */ \"use strict\"; var wrappedVar = %jsonContent; console.log(wrappedVar);\n", {jsonContent: jsonContent});
```
Hoisting this out to a module-scoped constant would improve the structure as well as reduce the line length

the-t-in-rtf replied:
Yup, done.

the-t-in-rtf committed:
GPII-3132: Added JSON dot files to lint checks and added trailing carriage returns.

the-t-in-rtf committed:
GPII-3132: Fixed file count in json-eslint task.

the-t-in-rtf commented:
I fixed the json-eslint file count, researching the mdjsonlint one now.

the-t-in-rtf committed:
GPII-3132: Upgraded gpii-grunt-mdjson-lint and standardised license details in package.json.

the-t-in-rtf commented:
I fixed the same problem with an unbracketed ternary operation in mdjsonlint, published a new dev release, and updated.  That file count is now there.

the-t-in-rtf committed:
GPII-3132: Moved wrapper string template to improve code readability.

the-t-in-rtf committed:
GPII-3132: Added an "other" category of sources so that we can check text files, etc. for carriage returns.

the-t-in-rtf committed:
GPII-3132: Updated tests to cover "other" category and add checks for lintspaces.

the-t-in-rtf commented:
@amb26, this is ready for further review.

the-t-in-rtf commented:
@amb26, I updated the [associated universal pull](https://github.com/GPII/universal/pull/623) to use the latest dev release, it's probably a good idea to take a look at the practical changes imposed by the new checks before we merge and cut a real release.

the-t-in-rtf committed:
GPII-3132: Limited lintspaces indentation to JSON files.

amb26 commented on package.json:
```diff
@@ -0,0 +1,46 @@
+        "fluid-grunt-eslint": "18.1.2",
+        "fluid-grunt-json5lint": "1.0.0",
+        "gpii-grunt-mdjson-lint": "1.0.0-dev.20180629T080410Z.02b3951",
+        "grunt": "~0.4.5",
```
Ancient, non-fixed grunt version - 1.0.3 will hopefully produce a few fewer warnings in install

the-t-in-rtf replied:
I think that was the grunt plugin init script's choice, it's fixed now.  I also updated the markdown linting plugin, which was way out of date.

amb26 commented on tasks/json-eslint.js:
```diff
@@ -0,0 +1,70 @@
+                fileCount++;
+                try {
+                    var jsonContent       = grunt.file.read(filepath);
+                    var wrappedContent = fluid.stringTemplate(wrapperTemplate, {jsonContent: jsonContent});
```
Wacky spacing here

the-t-in-rtf replied:
Fixed.

the-t-in-rtf committed:
GPII-3132: Updated dependencies based on output of `npm outdated`.

the-t-in-rtf committed:
GPII-3132: Fixed spacing in assignment block.

the-t-in-rtf commented:
@amb26, this should be ready for review again.

the-t-in-rtf commented:
Thanks, @amb26.  I have cut the release, updated the forward-facing version in master, and will shortly update the universal pull.
